proposals/0000-partial-field-behavior.md

```diff
@@ -0,0 +1,362 @@
+4. Tools that reflect on the shape of data will treat them differently (ie JSON parsing and subcommand parsing)
+
+
+### Other Languages
+
+How do other languages handle this?
+
+[TypeScript](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html) provides a compiler error if you use a partial field on a union type.
+However, if you refine the type using a `switch` statement, then you can use the fields, and this is guaranteed to be safe.
+Additionally, the compiler does not complain if you use a field that is present on every type.
+
+[Rust](https://doc.rust-lang.org/reference/items/enumerations.html) simply does not allow you to use dot notation on an enum with record fields, even if the field is shared among all types.
+To access named fields on an enum, you must use pattern matching.
+
+[F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions) also does not treat named fields on a union as accessors, and only uses them for construction and pattern matching.
+
+[OCaml](https://stackoverflow.com/questions/22367396/does-ocaml-have-record-syntax-in-disjoint-unions) does not allow you to define named fields in a union at all.
+
+[PureScript](https://jordanmartinez.github.io/purescript-jordans-reference-site/content/21-Hello-World/01-Prelude-ish/01-Basic-FP-Data-Types/01-Sum-and-Product-Types.html)'s sums-with-records are different - they contain a single value that is an anonymous record, instead of the constructor carrying a number of fields.
+
+Every language that has field labels in sum types allows you to use them in record construction and pattern matching.
+Contrasting these languages, then, we have this table:
+
+| Language   | Sum Labels?  | Field Access? | Runtime errors? |
+|------------|--------------|---------------|-----------------|
+| Haskell    |     Y        |       Y       |      Y          |
+| Rust       |     Y        |               |                 |      
+| F#         |     Y        |               |                 |
+| OCaml      |              |               |                 |
+| TypeScript |     Y        |       Y       |                 |
+| PureScript |              |               |                 |
+
+I don't know about you, but I don't like being the only language in the table with runtime errors about this.
+
+
+## Proposed Change Specification
+
+I propose we introduce a few new language extensions that allow us to customize the behavior of partial field selectors.
+The following are *options*, and are not intended to be a complete set that must remain unbroken.
+
+### `NoPartialFieldSelectors` + `PartialFieldSelectors`
+
+`PartialFieldSelectors` is the current behavior of GHC.
+So let's talk about `NoPartialFieldSelectors`:
+
+#### No selector function
+
+Under `NoPartialFieldSelectors`, we wouldn't get a partial function.
+
+```haskell
+data OptionRec a = None | Some { fromSome :: a }
```

Comment:
Do I understand it correctly that a field present in all variants would not be affected? Perhaps add such a field in the example to clarify this.